src/grpc/infer_handler.cc

```diff
@@ -785,6 +749,7 @@ ModelInferHandler::Process(InferHandler::State* state, bool rpc_ok)
 #endif  // TRITON_ENABLE_TRACING
 
     state->step_ = Steps::FINISH;
+  } else if (state->step_ == Steps::FINISH) {
     finished = true;
```

Comment:
I believe this introduces an extra state transition/iteration for every COMPLETE state. I think it makes more logical sense to have states transition to the FINISH state first if that's what it's intended for like you're doing here, but just double checking this is expected from your changes.

Reply:
Actually, please help clarify my understanding on something. Assume a state is sitting on completion queue in the COMPLETE step.

1. [correct if wrong] The tag loop `cq_->Next(&tag, &ok)` will **pop** the tag (state) off of the completion queue, and pass it to `Process(state)`. So at this point, the state is **no longer on the completion queue**?
2. Process(state) will hit the `else if state->step_ == Steps::COMPLETE)` condition, and simply update the `state->step_ = Steps::FINISH`, and then the `Process` function with exit with `finished == false`. 
3. [correct if wrong] There are no GRPC specific calls (read/write/finish) made, nor any manual `PutTaskBackToQueue` call to put this state back onto the queue. 
4. Since the state has been popped off the queue, not put back onto the queue, and `finished == false`, this state won't get cleaned up by the loop logic right? Is this a leak?
```cpp
  // This condition won't be true for a COMPLETE state with these changes if I 
  // understand correctly, described above, so state won't get released?
 if (!Process(state, ok)) {
        LOG_VERBOSE(1) << "Done for " << Name() << ", " << state->unique_id_;
        state->context_->EraseState(state);
        StateRelease(state);
  }
```

Reply:
Let's take a step further back. When the step of the state is moved to COMPLETE, the infer handler thread invokes ::grpc::ServerAsyncResponseWriter<inference::ModelInferResponse> 's [Finish()](https://github.com/triton-inference-server/server/blob/tanmayv-fix/src/grpc/infer_handler.cc#L947) call with the state. This marks the beginning of the sending gRPC response. 
Upon the completion of the sending response, there are 2 events that get posted on the completion queue.
1. Finish is complete the completion_queue places the state object passed to the Finish() call.
2. The AsyncNotifyWhenDone() places the StateWrapper object - pointing to the same state - to the completion queue. 

Now these two events can be popped from the completion queue in any order. Hence, we don't differentiate between the two. Whenever the first state is read from the Next call, we silently transition from COMPLETE->FINISH with `finished==false` and rely on the second event to get FINISH step state and return with `finish==true` so the state can be released.

Reply:
>Is this a leak? Will we run out of states in state bucket?
Most likely not. I have tested with all the popular cases and both the events are being properly delivered to us in CQ.